### PR TITLE
[webOS] Fix timezone reading

### PR DESCRIPTION
**Pull request by sundermann**
## Description

Timezone was not set correctly on startup on webOS. Some refactoring is done first to avoid duplicating code. The diff is a bit large because `clang format` required fixing up the header inclusion order and variable spacing

## Motivation and context

The webOS jailer only mounts `/etc/localtime` which is a symlink to `/var/luna/preferences/localtime`. `/var/luna/preferences/localtime ` however is another symlink to the actual zone info.

## How has this been tested?

webOS22/7

## What is the effect on users?

Fixes timezone initialization

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [x] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by sundermann** [comment]
It seems `std::filesystem` is not available until macOS 10.15 and `fstream` is not available until NDK 24.

**Comment by throwaway96** [comment]
Information for webOS 4.1.0 (MStar M3 SoC, using the YouTube jail as an example):

`/var/luna/preferences/timezone` is present in the jail, but it's `Universal` instead of the system timezone (`America/New_York`):
```
root@lg-m3[m3 4.1.0]:~# cat /var/palm/jail/youtube.leanback.v4/var/luna/preferences/timezone
Universal
```

`/etc/localtime` is a symlink to `/var/luna/preferences/localtime`:
```
root@lg-m3[m3 4.1.0]:~# readlink /var/palm/jail/youtube.leanback.v4/etc/localtime
/var/luna/preferences/localtime
```

`/var/luna/preferences/localtime` (symlink to correct zoneinfo) is present in the jail:
```
root@lg-m3[m3 4.1.0]:~# readlink /var/palm/jail/youtube.leanback.v4/var/luna/preferences/localtime
/usr/share/zoneinfo/America/New_York
```

It seems that `/usr/share/zoneinfo` is available from inside the jail, e.g.:
```
  File: /var/palm/jail/youtube.leanback.v4/usr/share/zoneinfo/America/New_York
  Size: 3545            Blocks: 7          IO Block: 1024   regular file
Device: b31dh/45853d    Inode: 61542       Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2018-09-17 22:29:54.000000000
Modify: 2018-09-17 22:29:54.000000000
Change: 2018-09-17 22:29:54.000000000
```

These are the relevant lines in `/etc/jail_native.conf` (although I'm not sure exactly which configuration file is used, or how it's chosen, these lines seem to be common to all):
```
# /var/luna/preferences
mkdir /var/luna
mkdir /var/luna/preferences
mount ro /var/luna/preferences

link /var/luna/preferences/localtime /etc/localtime
```

`/usr/share` (apparently responsible for the existence of `/usr/share/zoneinfo`, although they mount `/usr/share/fonts` explicitly) is also mounted by `jail_native.conf`:
```
mount ro /usr/share
```

These files are the same outside of the jail. It looks like `/var/luna/preferences/timezone` has the wrong timezone setting. That may be an artifact of the strange way my webOS boards are configured (missed a firstuse step?), but I can look into it on other boards/versions.

**Comment by sundermann** [comment]
Thank you for checking this on different webOS versions. As you suggested I have changed to reading the symlink path from `/var/luna/preferences/localtime` instead

File: xbmc/platform/posix/PosixTimezone.cpp [hunk 1]
```diff
@@ -195,44 +197,65 @@ void CPosixTimezone::SetTimezone(const std::string& timezoneName)
...
-       p = strrchr(timezoneName,'/');
-       *q = '/';
-       if (p)
-         p++;
-     }
-     return p;
-   }
+  // RHEL and maybe other distros make /etc/localtime a symlink
+  if (timezoneName.empty())
+    timezoneName = ReadFromLocaltime("/etc/localtime");
```
Comment (wsnipex) [inline xbmc/platform/posix/PosixTimezone.cpp:211]: this is how systemd configures the timezone, so it should work on any distro with systemd.
I'd make this the first option we try

Comment (sundermann) [reply xbmc/platform/posix/PosixTimezone.cpp:211]: `/etc/localtime` points to `/var/luna/preferences/localtime` on webOS.  So we'd retrieve `preferences/localtime` if it was tried first

Comment (wsnipex) [reply xbmc/platform/posix/PosixTimezone.cpp:211]: don't you already use readlink? And that's what was used before as well

Comment (sundermann) [reply xbmc/platform/posix/PosixTimezone.cpp:211]: Oh sorry, I meant `realpath`

Comment (wsnipex) [reply xbmc/platform/posix/PosixTimezone.cpp:211]: that should work as well. Quickly tested on ubuntu

Comment (sundermann) [reply xbmc/platform/posix/PosixTimezone.cpp:211]: Do you think it's better to change only to `realpath` and keep the rest of the code or also apply the refactoring as well and adapting that to use `realpath` too?

**Review by wsnipex — APPROVED** [review]
looks good to me

**Pull request merged**